piker/brokers/kucoin.py

```diff
@@ -0,0 +1,670 @@
+            ).digest()
+        )
+
+        passphrase = base64.b64encode(
+            hmac.new(
+                self._key_secret.encode('utf-8'),
+                self._key_passphrase.encode('utf-8'),
+                hashlib.sha256,
+            ).digest()
+        )
+
+        return {
+            'KC-API-SIGN': signature,
+            'KC-API-TIMESTAMP': str(pendulum.now().int_timestamp * 1000),
+            'KC-API-KEY': self._key_id,
+            'KC-API-PASSPHRASE': passphrase,
+            # XXX: Even if using the v1 api - this stays the same
+            'KC-API-KEY-VERSION': '2',
+        }
+
+    async def _request(
+        self,
+        action: Literal['POST', 'GET'],
+        endpoint: str,
+        api_v: str = 'v2',
+        headers: dict = {},
+    ) -> Any:
+        '''
+        Generic request wrapper for Kucoin API
+
+        '''
+        if self._config:
+            headers = self._gen_auth_req_headers(action, endpoint, api_v)
+
+        api_url = f'https://api.kucoin.com/api/{api_v}{endpoint}'
+
+        res = await asks.request(action, api_url, headers=headers)
+
+        if 'data' in res.json():
+            return res.json()['data']
+        else:
+            log.error(
+                f'Error making request to {api_url} -> {res.json()["msg"]}')
+            return res.json()['msg']
+
+    async def _get_ws_token(self, private: bool = False) -> tuple[str, int] | None:
+        '''
+        Fetch ws token needed for sub access:
+        https://docs.kucoin.com/#apply-connect-token
+        returns a token and the interval we must ping
+        the server at to keep the connection alive
+
+        '''
+        token_type = 'private' if private else 'public'
+        data: dict[str, Any] | None = await self._request(
+            'POST', f'/bullet-{token_type}', 'v1'
+        )
+
+        if data and 'token' in data:
+            # ping_interval is in ms
+            ping_interval: int = data['instanceServers'][0]['pingInterval']
+            return data['token'], ping_interval
+        elif data:
+            log.error(
+                f'Error making request for Kucoin ws token -> {data.json()["msg"]}'
+            )
+
+    async def _get_pairs(
+        self,
+    ) -> dict[str, KucoinMktPair]:
+        entries = await self._request('GET', '/symbols')
+        syms = {kucoin_sym_to_fqsn(item['name']): KucoinMktPair(**item) for item in entries}
+
+        log.info('Kucoin market pairs fetched')
+        return syms
+
+    async def cache_pairs(
+        self,
+        # normalize: bool = True,
+    ) -> dict[str, KucoinMktPair]:
+        '''
+        Get cached pairs and convert keyed symbols into fqsns if ya want
+
+        '''
+        if not self._pairs:
+            self._pairs = await self._get_pairs()
+
+        return self._pairs
+
+    async def search_symbols(
+        self,
+        pattern: str,
+        limit: int = 30,
+    ) -> dict[str, KucoinMktPair]:
+        data = await self._get_pairs()
```

Comment:
wouldn't it make more sense to use `.cache_pairs()` for this call so you aren't constantly reloading?